esx_service/vmdk_ops.py

```diff
@@ -342,51 +347,109 @@ def detachVMDK(vmdk_path, vm_uuid):
     return disk_detach(vmdk_path, vm)
 
 
-# Check existence (and creates if needed) the path
-def getVolPath(vm_config_path):
-    # The folder for Docker volumes is created on <vm_datastore>/DOCK_VOLS_DIR
-    # On ESX the path is always /vmfs/volume/datastore/... , so we can do this:
-    vm_datastore = "/".join(vm_config_path.split("/")[0:4])
-    path = os.path.join(vm_datastore, DOCK_VOLS_DIR)
-
+# Check existence (and creates if needed) the path for docker volume VMDKs
+def get_vol_path(datastore):
+    # The folder for Docker volumes is created on <datastore>/DOCK_VOLS_DIR
+    path = os.path.join("/vmfs/volumes", datastore, DOCK_VOLS_DIR)
+    
     if os.path.isdir(path):
         # If the path exists then return it as is.
         logging.debug("Found %s, returning", path)
         return path
-
+    
     # The osfs tools are usable for all datastores
     cmd = "{0} {1}".format(OSFS_MKDIR_CMD, path)
     rc, out = RunCommand(cmd)
-    if rc != 0:
-        logging.warning("Failed to create %s", path)
-    else:
+    if rc == 0:
         logging.info("Created %s", path)
         return path
-
+    
+    logging.warning("Failed to create %s", path)
     return None
 
 
-def getVmdkName(path, vol_name):
-    # form full name as <path-to-volumes>/<volname>.vmdk
-    return os.path.join(path, "%s.vmdk" % vol_name)
+def known_datastores():
+    """returns names of know datastores"""
+    return [i[0] for i in vmdk_utils.get_datastores()]
 
+def parse_vol_name(full_vol_name):
+    """
+    Parses volume[@datastore] and returns (volume, datastore)
+    On parse errors raises ValidationError with syntax explanation
+    """
+    groups = re.match("\A([\w\-.]+)@?(\w+)?$", full_vol_name)
+    if not groups:
+        raise ValidationError("Invalid syntax: '%s'. " \
+                           "Valid syntax is volume@datastore, where volume is " \
+                           "[a-zA-Z0-9_-. and datastore is [a-zA-Z0-9_]" % 
+                           full_vol_name)
+    return groups.groups()[0], groups.groups()[1]
+
+
+def get_full_vol_name(vmdk_name, datastore, vm_datastore):
+    """
+    Forms full volume name from vmdk file name an datastore as volume@datastore
+    For volumes on vm_datastore, just returns volume name
+    """
+    vol_name = vmdk_utils.strip_vmdk_extension(vmdk_name)
+    logging.debug("get_full_vol_name: %s %s %s", vmdk_name, datastore, vm_datastore)
+    if datastore == vm_datastore:
+        return vol_name
+    return "{0}@{1}".format(vol_name, datastore)
+
+
+# TBD - move to vmdk_utils
+def get_datastore_name(config_path):
+    """Returns datastore NAME in config_path (not url-name which may be used in path)"""
+    # path is always /vmfs/volumes/<datastore>/... , so extract datastore:
+    config_ds_name = config_path.split("/")[3]
+    ds_name = [x[0] for x in vmdk_utils.get_datastores() \
```

Comment:
it should not, but I felt it's more reliable this way :-)  - no harm in being paranoid